### ISSUE TITLE
Multi user (work-profile) support

### ISSUE THREAD

**Issue by TheMysticS** [d4rken-org/mystic-permissions#20]
If an app is installed in both the Main Profile and the Work Profile, only one of them is shown in the Apps tab.

In the attached screenshot, you'll see that the File Manager is showing up only once, as being installed by Shelter app. So this app is from the Work Profile.

I have the same app installed in the main profile too using SAI. It is not showing up.

[Screenshot_2022-07-17-20-15-12-95_ce34de6323cfaeb7a2880d42987990b1](https://user-images.githubusercontent.com/54448474/179403758-48e56efa-1bee-4033-b2e4-0dc961ea8142.jpg)

**Comment by d4rken** [comment]
Do both show in SD Maid?

**Comment by TheMysticS** [comment]
> Do both show in SD Maid?
>
It is not showing in SD Maid either. Only one app is showing up, and I'm not sure which one it is. I'm using SD Maid v5.4.0 Beta

**Comment by TheMysticS** [comment]
![Screenshot_2022-07-17-21-41-26-13_62f0e9ef2a35bdc7c62d39832d484c48](https://user-images.githubusercontent.com/54448474/179410873-a13b82b8-152d-4f03-924b-9374ba0eb7af.jpg)

**Comment by TheMysticS** [comment]
Should I open a new ticket for SD Maid too on this issue?

**Comment by d4rken** [comment]
Which app shows 2? How do you know there should be 2?

**Comment by TheMysticS** [comment]
> Which app shows 2? How do you know there should be 2?

How do I know? I'm the one who has installed the app in both profiles. And it is not just this app. There are other apps too that have been installed in both profiles. But only one installation is shown in both SD Maid and Permissions Overview.

Technically speaking, an app is installed only once in a device. Any other 'installation', be it in Work Profile or a Multi-User environment, is just creation of a Data folder for that user, and not a second installation as such.

But when using AppControl, all such 'installations' must show up seperately.

**Comment by d4rken** [comment]
I'm not sure if I would consider work profile support as in scope.
The app is installed for a specific user/profile, it makes sense that it shows the permissions and apps of that user.

I'm currently installing shelter to see what is possible. Not sure if root support is something I'd consider for this app.

**Comment by d4rken** [comment]
When installed in the main profile it sees all the apps in the main profile, and if installed in the work profile only the ones in the work profile.

[Screenshot from 2022-07-17 20-05-13](https://user-images.githubusercontent.com/1439229/179419000-d5c4e40d-92e7-4322-803b-d542ecf9e3f5.png)

As far as I can see, I could detect that an app has a work profile, but that's about it.

For anything else we'd need to dive much deeper and do the same thing as Shelter does (if this is even possible in parallel with Shelter). So listing apps from other profiles is out of scope.

I don't think this is a strong enough use-case to warrant adding support for using root.

**Comment by TheMysticS** [comment]
The app is seeing all the packages (apps) installed from the main profile. This includes all apps installed in the Work Profile, and also potentially apps installed in multi-users (I haven't tested Multi-Users yet).

1. It is only a matter of identifying the profile where the app is currently installed to add the badge (if applicable). I'm not sure if this requires root support since the app is already showing apps installed in the work profile.

2. I tried to uninstall the app using SD Maid, and this is what happened:

    • Without root, it was unable to delete the app. This means that the app shown in SD Maid AppControl is the one from Work Profile.

    • With root, SD Maid is completely uninstalling the app, i.e. it is gone from both profiles in just a single run.
4.

**Comment by d4rken** [comment]
> The app is seeing all the packages (apps) installed from the main profile. 

The app sees the packages that are visible in the profile it is in.

> This includes all apps installed in the Work Profile, and also potentially apps installed in multi-users (I haven't tested Multi-Users yet).

Work profile is just multi-user with a fancy name. I've checked the paths on a rooted device and the data for the main profile was `/data/user/0/` while for the work profile it was `/data/user/10/`.

> It is only a matter of identifying the profile where the app is currently installed to add the badge (if applicable). I'm not sure if this requires root support since the app is already showing apps installed in the work profile, including those that are installed in the Work Profile alone (which means those apps are not installed in the main profile).

" It is only a matter of " :grin: If there is no API for it, then without root it can't really be done. I've checked all APIs I could think of, and there was none that could give me this information. If you know more, please tell me.

> I tried to uninstall the app using SD Maid, and this is what happened:
> • Without root, it was unable to delete the app. This means that the app shown in SD Maid AppControl is the one from Work Profile.

SD Maid does very elaborate queries to get even apps that are not really installed (out of scope for this app). The app that was only in the work profile was still visible to SD Maid. It was not visible as "work profile app", just as "it's not really installed, but still somehow here". When SD Maid asked the system to uninstall it, the system probably said "what app?" as from the main profile's perspective it is not installed.

> • With root, SD Maid is completely uninstalling the app, i.e. it is gone from both profiles in just a single run.

Basically works like issuing `adb uninstall`.

**Comment by d4rken** [comment]
Reopen as we have a chance using the `LauncherApps` api

**Comment by dragonGod97677** [comment]
> Reopen as we have a chance using the `LauncherApps` api

Just a question, that API does exist on all Android version or just newer version?

**Comment by d4rken** [comment]
According to the docs, since API21:
https://developer.android.com/reference/android/content/pm/LauncherApps

Code is already in the current builds.

https://github.com/d4rken-org/mystic-permissions/blob/744d14ed6dfe79da1f5e75602045b915e8f4e00e/app/src/main/java/eu/darken/myperm/apps/core/AppRepo.kt#L75-L78

Should work on your Android version if it doesn't crash already.

**Comment by TheMysticS** [comment]
For apps installed in multiple profiles, can you not list all of them in the Apps Tab? Those from other profiles will have the briefcase badge.

There will also be a separate filter for such apps as discussed. I'll open a new ticket for it.

**Comment by TheMysticS** [comment]
You already have a way to find apps installed in other profiles, isn't it?

Can't the output of that be used to list these in the Apps Tab?

**Comment by d4rken** [comment]
We have different amounts of information depending what app it is or rather what source the data comes from.
The normal system API gives more information about an app than the API we use for apps in other profiles (e.g. we don't know permission status). So they are treated differently in code.

Currently it doesn't feel like listing them as extra entry in the list is useful enough to warrant the additional code complexity.

Can you elaborate on your use-case?
When I think about:
You look for a workplace app, you just find it, one entry, no need to think about which one click.
Any actions that can be done will be available in the details screens.
Actions may only affect the workplace version or the main version or both, so it makes sense to have them in one place too.
If we split up apps and each have their own list entry, then we also need extra details screens for each.

**Comment by TheMysticS** [comment]
The main use case for this is to be able to see (from the Main Profile) all apps installed in the device. 

Work Profile and Multi-Users (there is a distinct advantage of a Work Profile over Multi-Users and vice-versa, although they are simply new users 10, 11, 12, etc.) support independent installations, both via Store apps and via side-loading.

This app, being one about permissions and information in one place, would give the user all the info he needs in one place.
